src/main/scala/org/constellation/session/SessionTokenService.scala

```diff
@@ -0,0 +1,86 @@
+package org.constellation.session
+
+import cats.effect.Concurrent
+import cats.effect.concurrent.Ref
+import cats.implicits._
+import io.chrisdavenport.fuuid.FUUID
+import io.circe.generic.semiauto._
+import io.circe.{Decoder, Encoder}
+import org.constellation.session.SessionTokenService._
+import org.http4s.util.CaseInsensitiveString
+
+class SessionTokenService[F[_]: Concurrent] {
+
+  private[session] val peersTokens: Ref[F, Map[String, Token]] = Ref.unsafe(Map.empty)
+
+  private[session] val ownToken: Ref[F, Option[Token]] = Ref.unsafe(None)
+
+  def createAndSetNewOwnToken: F[Token] =
+    for {
+      token <- createToken
+      _ <- ownToken.modify(_ => (token.some, ()))
+    } yield token
+
+  def getOwnToken: F[Option[Token]] =
+    ownToken.get
+
+  def updatePeerToken(ip: String, token: Token): F[Unit] =
+    peersTokens.modify(tokens => (tokens + (ip -> token), ()))
+
+  def getPeerToken(ip: String): F[Option[Token]] =
+    peersTokens.get.map(_.get(ip))
+
+  def verifyPeerToken(ip: String, oHeaderToken: Option[Token]): F[TokenVerificationResult] =
+    for {
+      oPeerToken <- getPeerToken(ip)
+      verificationResult = SessionTokenService.compareTokens(oPeerToken, oHeaderToken)
+    } yield verificationResult
+
+  def clearOwnToken: F[Unit] =
+    ownToken.modify(_ => (None, ()))
+
+  def clearPeerToken(ip: String): F[Unit] =
+    peersTokens.modify(tokens => (tokens - ip, ()))
+
+  def clear: F[Unit] =
+    for {
+      _ <- clearOwnToken
+      _ <- peersTokens.modify(_ => (Map.empty, ()))
+    } yield ()
+
+  private[session] def createToken: F[Token] =
+    for {
+      uuid <- FUUID.randomFUUID
+      token = new Token(uuid.toString)
+    } yield token
+
+}
+
+object SessionTokenService {
+  class Token(val value: String) extends AnyVal
```

Comment:
it should be `case class Token(value: String) extends AnyVal`. no need to make it class nor `val value`

Reply:
Isn't case class heavier than class? For this purpose I think something less heavy is appropriate.

Reply:
Sounds to me like an overoptimisation. Even without `AnyVal` it has only one property so won't be heavy I guess. You can combine both actually.

Reply:
Sure that may be an overoptimisation. I remember reading that bytecode for simple class vs simple case class is couple times smaller and it stuck in my head as a drawback - but the thing is it doesn't necessarily mean it has any performance overhead. I would happily use case class here, as it's better for holding data :).